### ISSUE TITLE
RuntimeError: failed to get urandom

### ISSUE THREAD

**Issue by WolfAHeidrich** [rubygems/rubygems#5941]
### Describe the problem as clearly as you can
Tried to install Redmine following the steps on https://www.redmine.org/projects/redmine/wiki/RedmineInstall.

### Did you try upgrading rubygems & bundler?
Yes. "gem update --system" went fine. "bundle update --bundler" returned "This Bundle hasn't been installed yet. Run `bundle install` to update and install the bundled gems."

### Which command did you run?
bundle install --without development test

### What were you expecting to happen?
Installation of Redmine

### What actually happened?
--- ERROR REPORT TEMPLATE -------------------------------------------------------

```
RuntimeError: failed to get urandom
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/3.1/securerandom.rb:75:in `urandom'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/3.1/securerandom.rb:75:in `singleton class'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/3.1/securerandom.rb:42:in `<module:SecureRandom>'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/3.1/securerandom.rb:41:in `<top (required)>'
  <internal:/share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/rubygems/core_ext/kernel_require.rb>:85:in `require'
  <internal:/share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/rubygems/core_ext/kernel_require.rb>:85:in `require'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/fetcher.rb:5:in `<top (required)>'
  <internal:/share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/rubygems/core_ext/kernel_require.rb>:85:in `require'
  <internal:/share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/rubygems/core_ext/kernel_require.rb>:85:in `require'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/cli/install.rb:50:in `run'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/cli.rb:257:in `block in install'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/settings.rb:131:in `temporary'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/cli.rb:256:in `install'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/cli.rb:31:in `dispatch'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/cli.rb:25:in `start'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/gems/3.1/gems/bundler-2.3.22/exe/bundle:48:in `block in <main>'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/site_ruby/3.1/bundler/friendly_errors.rb:120:in `with_friendly_errors'
  /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/gems/3.1/gems/bundler-2.3.22/exe/bundle:36:in `<main>'

```

## Environment

```
Bundler       2.3.22
  Platforms   ruby, arm-linux-gnu
Ruby          3.1.2p20 (2022-04-12 revision 4491bb740a9506d76391ac44bb2fe6e483fec952) [arm-linux-gnu]
  Full Path   /opt/bin/ruby
  Config Dir  /opt/etc
RubyGems      3.3.22
  Gem Home    /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/gems/3.1
  Gem Path    /root/.gem/ruby/3.1:/share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/gems/3.1
  User Home   /root
  User Path   /root/.gem/ruby/3.1
  Bin Dir     /share/CE_CACHEDEV1_DATA/.qpkg/Entware/lib/ruby/gems/3.1/bin
OpenSSL
  Compiled    OpenSSL 1.1.1q  5 Jul 2022
  Loaded      OpenSSL 1.1.1q  5 Jul 2022
  Cert File   /opt/etc/ssl/cert.pem
  Cert Dir    /opt/etc/ssl/certs
Tools
  Git         2.34.3
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-09-07
Git SHA           44fb4c9ef5
Released Version  true
```

## Bundler settings

```
without
  Set for your local app (/share/CE_CACHEDEV1_DATA/homes/admin/redmine-5.0/.bundle/config): [:development, :test]
```

### Gemfile

```ruby
source 'https://rubygems.org'

ruby '>= 2.5.0', '< 3.2.0'
gem 'bundler', '>= 1.12.0'

gem 'rails', '6.1.7'
gem 'globalid', '~> 0.4.2' if Gem.ruby_version < Gem::Version.new('2.6.0')
gem 'rouge', '~> 3.28.0'
gem 'request_store', '~> 1.5.0'
gem 'mini_mime', '~> 1.1.0'
gem "actionpack-xml_parser"
gem 'roadie-rails', (Gem.ruby_version < Gem::Version.new('2.6.0') ? '~> 2.2.0' : '~> 3.0.0')
gem 'marcel'
gem "mail", "~> 2.7.1"
gem 'csv', '~> 3.2.0'
gem 'nokogiri', (Gem.ruby_version < Gem::Version.new('2.6.0') ? '~> 1.12.5' : '~> 1.13.4')
gem "rexml", require: false if Gem.ruby_version >= Gem::Version.new('3.0')
gem 'i18n', '~> 1.10.0'
gem "rbpdf", "~> 1.20.0"
gem 'addressable'
gem 'rubyzip', '~> 2.3.0'
gem 'net-smtp', '~> 0.3.0'
gem 'net-imap', '~> 0.2.2'
gem 'net-pop', '~> 0.1.1'
# Rails 6.1.6.1 does not work with Pysch 3.0.2, which is installed by default with Ruby 2.5. See https://github.com/rails/rails/issues/45590
gem 'psych', '>= 3.1.0' if Gem.ruby_version < Gem::Version.new('2.6.0')

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :x64_mingw, :mswin]

# TOTP-based 2-factor authentication
gem 'rotp', '>= 5.0.0'
gem 'rqrcode'

# Optional gem for LDAP authentication
group :ldap do
  gem 'net-ldap', '~> 0.17.0'
end

# Optional gem for exporting the gantt to a PNG file
group :minimagick do
  gem 'mini_magick', '~> 4.11.0'
end

# Optional Markdown support, not for JRuby
# ToDo: Remove common_mark group when common_mark is decoupled from markdown. See defect (#36892) for more details.
gem 'redcarpet', '~> 3.5.1', groups: [:markdown, :common_mark]

# Optional CommonMark support, not for JRuby
group :common_mark do
  gem "html-pipeline", "~> 2.13.2"
  gem "commonmarker", (Gem.ruby_version < Gem::Version.new('2.6.0') ? '0.21.0' : '0.23.4')
  gem "sanitize", "~> 6.0"
  gem 'deckar01-task_list', '2.3.2'
end

# Include database gems for the adapters found in the database
# configuration file
require 'erb'
require 'yaml'
database_file = File.join(File.dirname(__FILE__), "config/database.yml")
if File.exist?(database_file)
  yaml_config = ERB.new(IO.read(database_file)).result
  database_config = YAML.respond_to?(:unsafe_load) ? YAML.unsafe_load(yaml_config) : YAML.load(yaml_config)
  adapters = database_config.values.map {|c| c['adapter']}.compact.uniq
  if adapters.any?
    adapters.each do |adapter|
      case adapter
      when 'mysql2'
gem "mysql2", "~> 0.5.0", :platforms => [:mri, :mingw, :x64_mingw]
      when /postgresql/
gem "pg", "~> 1.2.2", :platforms => [:mri, :mingw, :x64_mingw]
      when /sqlite3/
gem "sqlite3", "~> 1.4.0", :platforms => [:mri, :mingw, :x64_mingw]
      when /sqlserver/
gem "tiny_tds", "~> 2.1.2", :platforms => [:mri, :mingw, :x64_mingw]
gem "activerecord-sqlserver-adapter", "~> 6.1.0", :platforms => [:mri, :mingw, :x64_mingw]
      else
warn("Unknown database adapter `#{adapter}` found in config/database.yml, use Gemfile.local to load your own database gems")
      end
    end
  else
    warn("No adapter found in config/database.yml, please configure it first")
  end
else
  warn("Please configure your config/database.yml first")
end

group :development do
  gem 'listen', '~> 3.3'
  gem "yard"
end

group :test do
  gem "rails-dom-testing"
  gem 'mocha', '>= 1.4.0'
  gem 'simplecov', '~> 0.21.2', :require => false
  gem "ffi", platforms: [:mingw, :x64_mingw, :mswin]
  # For running system tests
  gem 'puma'
  gem 'capybara', (Gem.ruby_version < Gem::Version.new('2.6.0') ? '~> 3.35.3' : '~> 3.36.0')
  gem "selenium-webdriver", "~> 3.142.7"
  gem 'webdrivers', '4.6.1', require: false
  # RuboCop
  gem 'rubocop', '~> 1.26.0'
  gem 'rubocop-performance', '~> 1.13.0'
  gem 'rubocop-rails', '~> 2.14.0'
end

local_gemfile = File.join(File.dirname(__FILE__), "Gemfile.local")
if File.exist?(local_gemfile)
  eval_gemfile local_gemfile
end

# Load plugins' Gemfiles
Dir.glob File.expand_path("../plugins/*/{Gemfile,PluginGemfile}", __FILE__) do |file|
  eval_gemfile file
end
```

### Gemfile.lock

```
<No /share/CE_CACHEDEV1_DATA/homes/admin/redmine-5.0/Gemfile.lock found>
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=failed+to+get+urandom&type=Issues

If there aren't any reports for this error yet, please fill in the new issue form located at https://github.com/rubygems/rubygems/issues/new?labels=Bundler&template=bundler-related-issue.md, and copy and paste the report template above in there.

**Comment by noraj** [comment]
```
RuntimeError: failed to get urandom
  /usr/lib/ruby/3.0.0/securerandom.rb:102:in `urandom'
  /usr/lib/ruby/3.0.0/securerandom.rb:102:in `singleton class'
  /usr/lib/ruby/3.0.0/securerandom.rb:69:in `<module:SecureRandom>'
  /usr/lib/ruby/3.0.0/securerandom.rb:68:in `<top (required)>'
  <internal:/usr/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
  <internal:/usr/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/fetcher.rb:5:in `<top (required)>'
  <internal:/usr/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
  <internal:/usr/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli/install.rb:50:in `run'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:257:in `block in install'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/settings.rb:131:in `temporary'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:256:in `install'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:31:in `dispatch'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:25:in `start'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/exe/bundle:48:in `block in <top (required)>'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
  /usr/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/exe/bundle:36:in `<top (required)>'
  /usr/bin/bundle:25:in `load'
  /usr/bin/bundle:25:in `<main>'

```

## Environment

```
Bundler       2.3.22
  Platforms   ruby, x86_64-linux
Ruby          3.0.4p208 (2022-04-12 revision 3fa771ddedac25560be57f4055f1767e6c810f58) [x86_64-linux]
  Full Path   /usr/bin/ruby
  Config Dir  /etc
RubyGems      3.3.19
  Gem Home    /usr/lib/ruby/gems/3.0.0
  Gem Path    /root/.local/share/gem/ruby/3.0.0:/usr/lib/ruby/gems/3.0.0
  User Home   /root
  User Path   /root/.local/share/gem/ruby/3.0.0
  Bin Dir     /usr/bin
OpenSSL
  Compiled    OpenSSL 1.1.1q  5 Jul 2022
  Loaded      OpenSSL 1.1.1q  5 Jul 2022
  Cert File   /etc/ssl/cert.pem
  Cert Dir    /etc/ssl/certs
Tools
  Git         2.37.3
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-09-12
Git SHA           44fb4c9ef
Released Version  true
```

## Bundler settings

```
build.nokogiri
  Set for your local app (/usr/share/cmsscanner/.bundle/config): "--use-system-libraries"
path
  Set for your local app (/usr/share/cmsscanner/.bundle/config): "vendor/bundle"
```

## Gemfile

### Gemfile

```ruby
# frozen_string_literal: true

source 'https://rubygems.org'
gemspec
```

### Gemfile.lock

```
<No /usr/share/cmsscanner/Gemfile.lock found>
```

## Gemspecs

### cms_scanner.gemspec

```ruby
lib = File.expand_path('../lib', __FILE__)
$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)

require 'cms_scanner/version'

Gem::Specification.new do |s|
  s.name                  = 'cms_scanner'
  s.version               = CMSScanner::VERSION
  s.platform              = Gem::Platform::RUBY
  s.required_ruby_version = '>= 2.5'
  s.authors               = ['WPScanTeam']
  s.email                 = ['contact@wpscan.com']
  s.summary               = 'CMS Scanner Framework (experimental)'
  s.description           = 'Framework to provide an easy way to implement CMS Scanners'
  s.homepage              = 'https://github.com/wpscanteam/CMSScanner'
  s.license               = 'MIT'

  s.files                 = Dir.glob('lib/**/*') + Dir.glob('app/**/*') + %w[LICENSE README.md]
  s.test_files            = []
  s.require_paths         = ['lib']

  s.add_dependency 'ethon', '>= 0.14', '< 0.16' # https://github.com/typhoeus/ethon/issues/185
  s.add_dependency 'get_process_mem', '~> 0.2.5'
  s.add_dependency 'nokogiri', '>= 1.11.4', '< 1.14.0'
  s.add_dependency 'opt_parse_validator', '~> 1.9.5'
  s.add_dependency 'public_suffix', '>= 4.0.3', '< 5.1.0'
  s.add_dependency 'ruby-progressbar', '>= 1.10', '< 1.12'
  s.add_dependency 'typhoeus', '>= 1.3', '< 1.5'
  s.add_dependency 'xmlrpc', '~> 0.3'
  s.add_dependency 'yajl-ruby', '~> 1.4.1' # Better JSON parser regarding memory usage

  s.add_dependency 'sys-proctable', '~> 1.2.2' # Required by get_process_mem for Windows OS.

  s.add_development_dependency 'bundler',             '>= 1.6'
  s.add_development_dependency 'rake',                '~> 13.0'
  s.add_development_dependency 'rspec',               '~> 3.11.0'
  s.add_development_dependency 'rspec-its',           '~> 1.3.0'
  s.add_development_dependency 'rubocop',             '~> 1.26.0'
  s.add_development_dependency 'rubocop-performance', '~> 1.13.0'
  s.add_development_dependency 'simplecov',           '~> 0.21.0'
  s.add_development_dependency 'simplecov-lcov',      '~> 0.8.0'
  s.add_development_dependency 'webmock',             '~> 3.17.0'
end
```

**Comment by flavorjones** [comment]
This does not appear to be a bundler issue. You may want to google a bit, and some more information is available at https://bugs.ruby-lang.org/issues/14716

**Comment by deivid-rodriguez** [comment]
Could you follow that line of investigation and bisect which Bundler version (if any) introduced this? Also, does this also happen if you lazily load `securerandom`:

```diff
diff --git a/bundler/lib/bundler/fetcher.rb b/bundler/lib/bundler/fetcher.rb
index e399a50cfd..66f77e028a 100644
--- a/bundler/lib/bundler/fetcher.rb
+++ b/bundler/lib/bundler/fetcher.rb
@@ -2,7 +2,6 @@
 
 require_relative "vendored_persistent"
 require "cgi"
-require "securerandom"
 require "zlib"
 require "rubygems/request"
 
@@ -193,6 +192,7 @@ def user_agent
         agent << " ci/#{cis.join(",")}" if cis.any?
 
         # add a random ID so we can consolidate runs server-side
+        require "securerandom"
         agent << " " << SecureRandom.hex(8)
 
         # add any user agent strings set in the config
```

**Comment by noraj** [comment]
I wasn't able to reproduce it manually in docker.

**Comment by noraj** [comment]
Entering into the nspawn container and running the commands manually works.

```
$ bundle config build.nokogiri --use-system-libraries
$ bundle config set --local path 'vendor/bundle'
$ bundle install
```

I encountered the same behavior with many gems that where previously building correctly (https://github.com/BlackArch/blackarch/issues/3519).

It seems to be specific to some very special environments under some special circumstances.

I found other older references of @WolfAHeidrich issue there https://github.com/docker-library/ruby/issues/383 and there https://github.com/docker-library/redmine/issues/264. But the solution is a bit different since this person was using the kernel `3.16.0` on is host while in my case both host and container use `5.19.13` and can't reproduce on docker `ruby:3.1-slim-bullseye`.

Container (nspawn):

```
$ ulimit -a
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) 0
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 22990
max locked memory           (kbytes, -l) 8192
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1024
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) 22990
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited
```

Host:

```
$ ulimit -a
-t: cpu time (seconds)              unlimited
-f: file size (blocks)              unlimited
-d: data seg size (kbytes)          unlimited
-s: stack size (kbytes)             8192
-c: core file size (blocks)         unlimited
-m: resident set size (kbytes)      unlimited
-u: processes                       22990
-n: file descriptors                1024
-l: locked-in-memory size (kbytes)  8192
-v: address space (kbytes)          unlimited
-x: file locks                      unlimited
-i: pending signals                 22990
-q: bytes in POSIX msg queues       819200
-e: max nice                        0
-r: max rt priority                 0
-N 15: rt cpu time (microseconds)   unlimited
```

**Comment by Andersenas** [comment]
Hello, 

I encountered a similar error during a `bundle install` on a Debian 11 VM (Ruby 3.31, Bundler 2.4.22):

```
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    failed to get urandom

Gem files will remain installed in bundle/ruby/3.3.0/gems/racc-1.7.3 for inspection.
Results logged to bundle/ruby/3.3.0/extensions/x86_64-linux/3.3.0/racc-1.7.3/gem_make.out

  /usr/local/lib/ruby/3.3.0/tmpdir.rb:133:in `urandom'
```

Following a colleague's advice, I investigated the Ruby 3 source code to understand the issue. The error message in Ruby:
https://github.com/ruby/ruby/blob/dd578cf2f134e915abdf41393fc08f29640fb9c9/random.c#L773
does not fully explain why the method execution might fail, as it can occur for at least three different reasons.

I discovered that the problem lay in how the `fill_random_bytes_urandom` method was built during Ruby's compilation:
https://github.com/ruby/ruby/blob/dd578cf2f134e915abdf41393fc08f29640fb9c9/random.c#L443

The final implementation of this method depends on your system, specifically on variables set in the `include/ruby/internal/config.h` file, which is generated after running `./configure ...`.

In my case, the variable `#define HAVE_GETENTROPY 1` was set, indicating that cryptographically secure random data should be obtained via `getentropy()`.

Since `getentropy()` has been available in Linux since kernel version 3.17, the "failed to get urandom" error occurred on my VM because the Linux kernel version was not updated when I upgraded to Debian 11, remaining at:

```
# uname -r
3.16.0-6-amd64
```

After updating the kernel version, the error no longer appeared.

You might have the same or a similar problem. If everything is fine with your Linux system, I suggest examining the `random.c` file for more insights:
https://github.com/ruby/ruby/blob/dd578cf2f134e915abdf41393fc08f29640fb9c9/random.c#L443

I hope this helps you resolve or get closer to resolving your issue. :)
